luarules/gadgets/unit_inherit_creation_xp.lua

```diff
@@ -82,70 +79,69 @@ local lastRunFrame = 0
+						childtype = "DRONE",
+					}
 				end
-				if childrenWithParents[unitID] then
-					parentID = childrenWithParents[unitID].parentunitid
-					local oldXP = oldChildXPValues[unitID] or 0
-					local newXP = Spring.GetUnitExperience(unitID) or 0
-					local parentXP = Spring.GetUnitExperience(parentID) or 0
-					local multiplier = childrenWithParents[unitID].parentxpmultiplier
-					local gainedXP = (parentXP*10000000 + ((newXP-oldXP)*10000000*multiplier))/10000000 --10000000 is to prevent very small numbers being lost during calculation
-					oldChildXPValues[unitID] = newXP
-					Spring.SetUnitExperience(parentID, gainedXP)
+			end
+			if Spring.GetUnitRulesParam(unitID, "parent_unit_id") then --estabalishes unit_explosion_spawner parenthood
+				parentID = Spring.GetUnitRulesParam(unitID, "parent_unit_id")
+				parentDefID = Spring.GetUnitDefID(parentID)
+				if parentsInheritXP[parentDefID] ~= nil and string.find(parentsInheritXP[parentDefID], "BOTCANNON") then
+					childrenWithParents[unitID] = {
+						unitid = unitID,
+						parentunitid = parentID,
+						parentxpmultiplier = calculatePowerDiffXP(unitID, parentID),
+						childinheritsXP = childrenInheritXP[unitDefID],
+						childtype = "BOTCANNON",
+					}
+				end
+			end
+			if childrenWithParents[unitID] then
+				parentID = childrenWithParents[unitID].parentunitid --sets parentID if it's not already set
+				parentDefID = Spring.GetUnitDefID(parentID) -- gets the parentDefID
+			end
+			if parentID ~= nil and childrenInheritXP[parentDefID] and childrenWithParents[unitID] then --if the parent has the unitdef, set childxp to parent xp.
+				local parentTypes = childrenInheritXP[parentDefID]
+				if string.find(parentTypes, childrenWithParents[unitID].childtype) then -- if child is correcty type, set xp
+					local parentXP = Spring.GetUnitExperience(parentID)
+					Spring.SetUnitExperience(unitID, parentXP)
+					oldChildXPValues[unitID] = parentXP	--add parent xp to the oldxp value to exclude it from inheritance
 				end
 			end
+			if inheritChildrenXP[unitDefID] then -- if parent inherits xp, then add to list of units to inherit xp
+				xpGainParents[unitID] = true
+			end
+			if not xpGainParents[unitID] and not childrenWithParents[unitID] then --if not a parent or a child of a parent who inherits xp, add to ignore list
+				ignoreList[unitID] = true
+			end
+			initializeList[unitID] = nil -- this concludes innitialization
+		end
+		
+		
+		for unitID, value in pairs(childrenWithParents) do
+			parentID = childrenWithParents[unitID].parentunitid
+			local oldXP = oldChildXPValues[unitID] or 0
+			local newXP = Spring.GetUnitExperience(unitID) or 0
+			local parentXP = Spring.GetUnitExperience(parentID) or 0
+			local multiplier = childrenWithParents[unitID].parentxpmultiplier
+			local gainedXP = (parentXP*10000000 + ((newXP-oldXP)*10000000*multiplier))/10000000 --10000000 is to prevent very small numbers being lost during calculation
```

Comment:
local gainedXP = (parentXP*10000000 + ((newXP-oldXP)*10000000*multiplier))/10000000 --10000000 is to prevent very small numbers being lost during calculation
Lua only supports floats as its number type. How does this help avoid losing precision?

Reply:
XP gain values were some ridiculously small fractional number. Something like .00000045 XP gained was pretty typical. There was a limit to the number of digits, maybe 12 IIRC?

So when I would start performing math on it, especially division operations, I would lose digits and get inconsistent and anomalous results.

Reply:
The code above, to my best attempt at understanding it, does nothing to 'fix' float underflow. 
The max precision you can expect is the standard float32.

Reply:
Without multiplying by 10 million or whatever I did up there, small amounts of inherited experience would cause the parents to actually *lose* XP. 
Either that, or it would gain significantly less XP than it should.

I believe it's because the XP gained is right at the border of the underflow. So when I applied the "multiplier" in that formula, which is itself a fraction, I would lose a digit or two.

Oh yeah, and that multiplier is locally defined by the XP differential function. 
The actual number it spits out for the multiplier is usually something like .000625 because there is usually a significant power differential between parent and child.

Reply:
I had at one point had to print every step of the calculation to see what was going on.

Perhaps engine reduces resolution to conserve resources due to the sheer number of XP calculations every frame.

If you don't believe me, I will insert a bunch of prints again and reverify that is the case in roughly 15 hours.

Reply:
Keep in mind that prints might limit/round the values far more than the variable does.

Reply:
Yeah, Beherith is right. I printed every step and couldn't see a difference. Yet, this fixed it when I originally wrote it. I'm just gonna hit the "believe it" button and remove silly extra 0's.